compiler/sem/sempass2.nim

```diff
@@ -1621,29 +1621,27 @@ proc detectCapture(owner, top: PSym, n: PNode, marker: var IntSet): PNode =
         # unwind
         return
 
-proc isCompileTimeProc2*(s: PSym): bool =
-  ## Tests if `s` is a compile-time procedure, but compared to
-  ## ``isCompileTimeProc``, also considers where `s` is defined. In the
-  ## following example:
-  ##
-  ## .. code-block:: nim
-  ##
-  ##   proc a() {.compileTime.} =
-  ##     proc b() =
-  ##       discard
-  ##
-  ## ``b`` is also a procedure that can only be run at compile-time, but it
-  ## doesn't have the ``sfCompileTime`` flag set
-  # TODO: propagate the ``sfCompileTime`` flag downwards instead (likely in
-  #       sempass2) and use ``isCompileTimeProc`` directly
-  # FIXME: procedures defined inside ``static`` blocks are not detected
-  var s = s
-  while s.kind != skModule:
-    if isCompileTimeProc(s):
-      return true
+proc canCaptureFrom*(captor, target: PSym): bool =
+  ## Tests if the `captor` routine is allowed to capture a local from `target`,
+  ## taking the compile-time/run-time boundary into account:
+  ## - attempting to capture a local defined outside an inner macro from inside
+  ##   the macro is illegal
+  ## - closing over a local defined inside a compile-time-only routine from a
+  ##   routine than can also be used at run-time is only valid if the chain of
+  ##   enclosing routines leading up to `target` are all compile-time-only
+  template isCompTimeOnly(s: PSym): bool =
+    sfCompileTime in s.flags
+
+  var s = captor
+  while s != target:
+    if s.kind == skMacro or (s != captor and
+       isCompTimeOnly(s) != isCompTimeOnly(target)):
+      return false
     s = s.skipGenericOwner
 
-  result = false
+  # it's not allowed for a compile-time-only routine to close over a run-time
+  # location
+  result = not (isCompTimeOnly(captor) and not isCompTimeOnly(target))
```

Comment:
I had a bit too much fun, but does this flow better?

```nim
result = not captor.isCompTimeOnly or target.isCompTimeOnly
  var s = captor
  while result and s != target: # does this need to check for nil?
    result =
      s.kind != skMacro and
      (s == captor or        # for captor ensure not macro
       s.isCompTimeOnly == target.isCompTimeOnly) # ensure same time
    s = s.skipGenericOwner
```

Reply:
Pretty sure I did all the De Morgan's correctly. 🤞🏽

Reply:
It does flow a lot better -- thanks for the suggestion.